bin-prot/tests/layouts.rs

```diff
@@ -165,37 +191,125 @@ fn test_option_rule() {
 
     let example_none = vec![0x00]; // None
 
-    let mut de = Deserializer::from_reader_with_layout(example_none.as_slice(), rule.clone());
+    let mut de =
+        Deserializer::from_reader_with_layout(Cursor::new(example_none.as_slice()), rule.clone());
     let result: Value = Deserialize::deserialize(&mut de).expect("Failed to deserialize");
     println!("{:?}", result);
     assert_eq!(result, Value::Option(None));
 
+    test_roundtrip(&result, &example_none);
+
     let example_some = vec![0x01, 0x07]; // Some(7)
 
-    let mut de = Deserializer::from_reader_with_layout(example_some.as_slice(), rule);
+    let mut de = Deserializer::from_reader_with_layout(Cursor::new(example_some.as_slice()), rule);
     let result: Value = Deserialize::deserialize(&mut de).expect("Failed to deserialize");
 
-    assert_eq!(result, Value::Option(Some(Box::new(Value::Int(0x07)))))
+    assert_eq!(result, Value::Option(Some(Box::new(Value::Int(0x07)))));
+    test_roundtrip(&result, &example_some);
+}
+
+const MULTIPLE_CTOR_ARG_SUM_RULE: &str = r#"
+[
+  "Sum",
+  [
+    {
+      "ctor_name": "one",
+      "index": 0,
+      "ctor_args": [
+        [
+          "Record",
+          [
+            { "field_name": "first", "field_rule": ["Int"] }
+          ]
+        ],
+        [
+          "Record",
+          [
+            { "field_name": "second", "field_rule": ["Int"] }
+          ]
+        ]
+      ]
+    }
+  ]
+]
+"#;
+
+#[test]
+fn test_multiple_ctor_arg_sum_rule() {
+    let rule: BinProtRule = serde_json::from_str(MULTIPLE_CTOR_ARG_SUM_RULE).unwrap();
+    let example = vec![0x00, 0x05, 0x06]; // One({ first: 5 }, { second: 6})
+
+    let mut de = Deserializer::from_reader_with_layout(Cursor::new(example.as_slice()), rule);
```

Comment:
nit: can remove the Cursor wrapper as https://doc.rust-lang.org/std/io/trait.Read.html#impl-Read-2

Reply:
While it can be safely removed I think I might leave it but only in the test code. It was useful for debugging since `Cursor` implements `Seek` so you can figure out exactly which input byte it is failing at. It isn't used currently but if we add it back it in would be useful to have in the tests.